File: frontend/src/types/event.ts
export interface EventBase {
  id: string
  clientId: string
  lamport: number
  timestamp: string
  entityId: string
  type: string
  payload: object
}

export type AppEvent =
  | (EventBase & { type: 'ShopCreated'; payload: { name: string; color: string } })
  | (EventBase & { type: 'ShopRenamed'; payload: { name: string } })
  | (EventBase & { type: 'ShopColorChanged'; payload: { color: string } })
  | (EventBase & { type: 'ShopSoftDeleted'; payload: { deletedAt: string } })
  | (EventBase & { type: 'TagCreated'; payload: { name: string } })
  | (EventBase & { type: 'TagDeleted'; payload: {} })
  | (EventBase & { type: 'ItemCreated'; payload: { name: string; unit?: string; defaultQuantity?: number; description?: string; notes?: string } })
  | (EventBase & { type: 'ItemUpdated'; payload: { name?: string; unit?: string; defaultQuantity?: number; description?: string; notes?: string } })
  | (EventBase & { type: 'ItemSoftDeleted'; payload: { deletedAt: string } })
  | (EventBase & { type: 'ShopAssignedToItem'; payload: { shopId: string } })
  | (EventBase & { type: 'ShopRemovedFromItem'; payload: { shopId: string } })
  | (EventBase & { type: 'TagAssignedToItem'; payload: { tagId: string } })
  | (EventBase & { type: 'TagRemovedFromItem'; payload: { tagId: string } })
  | (EventBase & { type: 'ListCreated'; payload: { name: string } })
  | (EventBase & { type: 'ListRenamed'; payload: { name: string } })
  | (EventBase & { type: 'ListArchived'; payload: { archivedAt: string } })
  | (EventBase & { type: 'ListDeleted'; payload: { deletedAt: string } })
  | (EventBase & { type: 'ListItemAdded'; payload: { listId: string; itemId: string; state: 'active' | 'bought'; quantity?: number; unit?: string; notes?: string } })
  | (EventBase & { type: 'ListItemStateChanged'; payload: { state: 'active' | 'bought' } })
  | (EventBase & { type: 'ListItemQuantityChanged'; payload: { quantity: number; unit?: string } })
  | (EventBase & { type: 'ListItemRemoved'; payload: {} })
  | (EventBase & { type: 'ShopSkippedForListItem'; payload: { shopId: string } })
  | (EventBase & { type: 'ShopSkipCleared'; payload: { shopId: string } })
  | (EventBase & { type: 'ShoppingSessionStarted'; payload: { listId: string; shopId: string } })
  | (EventBase & { type: 'SessionItemBought'; payload: { itemId: string; quantity?: number; unit?: string } })
  | (EventBase & { type: 'SessionItemSkipped'; payload: { itemId: string } })

File: backend/models/event.go
package models

import (
	"encoding/json"
	"fmt"
)

// Event type constants mirroring frontend/src/types/event.ts (AppEvent union).
const (
	EventShopCreated             = "ShopCreated"
	EventShopRenamed             = "ShopRenamed"
	EventShopColorChanged        = "ShopColorChanged"
	EventShopSoftDeleted         = "ShopSoftDeleted"
	EventTagCreated              = "TagCreated"
	EventTagDeleted              = "TagDeleted"
	EventItemCreated             = "ItemCreated"
	EventItemUpdated             = "ItemUpdated"
	EventItemSoftDeleted         = "ItemSoftDeleted"
	EventShopAssignedToItem      = "ShopAssignedToItem"
	EventShopRemovedFromItem     = "ShopRemovedFromItem"
	EventTagAssignedToItem       = "TagAssignedToItem"
	EventTagRemovedFromItem      = "TagRemovedFromItem"
	EventListCreated             = "ListCreated"
	EventListRenamed             = "ListRenamed"
	EventListArchived            = "ListArchived"
	EventListUnarchived          = "ListUnarchived"
	EventListDeleted             = "ListDeleted"
	EventListItemAdded           = "ListItemAdded"
	EventListItemStateChanged    = "ListItemStateChanged"
	EventListItemQuantityChanged = "ListItemQuantityChanged"
	EventListItemRemoved         = "ListItemRemoved"
	EventShopSkippedForListItem  = "ShopSkippedForListItem"
	EventShopSkipCleared         = "ShopSkipCleared"
	EventShoppingSessionStarted  = "ShoppingSessionStarted"
	EventSessionItemBought       = "SessionItemBought"
	EventSessionItemSkipped      = "SessionItemSkipped"
	EventBugReported             = "BugReported"
)

// Event is the wire format of an append-only domain event, mirroring the
// AppEvent shape the frontend emits.
type Event struct {
	ID        string          `json:"id"`
	ClientID  string          `json:"clientId"`
	Lamport   int             `json:"lamport"`
	Timestamp string          `json:"timestamp"`
	EntityID  string          `json:"entityId"`
	Type      string          `json:"type"`
	Payload   json.RawMessage `json:"payload"`
}

func IsValidEventType(t string) bool {
	switch t {
	case EventShopCreated,
		EventShopRenamed,
		EventShopColorChanged,
		EventShopSoftDeleted,
		EventTagCreated,
		EventTagDeleted,
		EventItemCreated,
		EventItemUpdated,
		EventItemSoftDeleted,
		EventShopAssignedToItem,
		EventShopRemovedFromItem,
		EventTagAssignedToItem,
		EventTagRemovedFromItem,
		EventListCreated,
		EventListRenamed,
		EventListArchived,
		EventListUnarchived,
		EventListDeleted,
		EventListItemAdded,
		EventListItemStateChanged,
		EventListItemQuantityChanged,
		EventListItemRemoved,
		EventShopSkippedForListItem,
		EventShopSkipCleared,
		EventShoppingSessionStarted,
		EventSessionItemBought,
		EventSessionItemSkipped,
		EventBugReported:
		return true
	}
	return false
}

func (e Event) Validate() error {
	if e.ID == "" {
		return fmt.Errorf("event: missing id")
	}
	if e.Type == "" {
		return fmt.Errorf("event: missing type")
	}
	if !IsValidEventType(e.Type) {
		return fmt.Errorf("event: unknown type %q", e.Type)
	}
	if e.EntityID == "" {
		return fmt.Errorf("event: missing entityId")
	}
	if e.Lamport < 0 {
		return fmt.Errorf("event: negative lamport %d", e.Lamport)
	}
	if e.Payload != nil && !json.Valid(e.Payload) {
		return fmt.Errorf("event: invalid payload JSON")
	}
	return nil
}
